lib/hiera/interpolate.rb

```diff
@@ -59,9 +64,19 @@ def do_interpolation(data, scope, extra_data, context)
     end
     private :do_interpolation
 
-    def get_interpolation_method_and_key(data)
+    def get_interpolation_method_and_key(data, is_config)
       if (match = data.match(METHOD_INTERPOLATION))
-        case match[1]
+        key = match[1]
+        if(is_config)
+          if key == 'hiera'
+            # Letting this pass while parsing the config is very likely to result in another attempt to parse
+            # the config and hence, endless recursion.
+            raise Hiera::InterpolationInvalidValue, "Cannot use interpolation method 'hiera' in hiera configuration file"
```

Comment:
I've tried but not succeeded in creating such a configuration (tested among other things the example in the JIRA). I don't see how it can succeed given that any hiera lookup will call datasources which in turn will call back to interpolate.